runtime/doc/eval.txt

```diff
@@ -5266,5 +5275,113 @@ Usage: >vim
+Vim chooses which one to use based on 'clipmethod'.  Despite the name, it does
+not use the 'clipboard' option and should be treated separate from the
+clipboard functionality.  It essentially overrides the existing behaviour of
+the clipboard registers.
+
+						*clipboard-providers-no-clipboard*
+If the |+clipboard| feature is not enabled, then the "+" and "*" registers
+will not be enabled/available unless |v:clipmethod| is set to a provider.
+
+						*clipboard-providers-plus*
+If on a platform that only has the "*" register, then the "+" register will
+only be available when |v:clipmethod| is set to a provider.  If you want to
+check if the "+" is available for use, it can be checked with: >
+	if has('unnamedplus')
+<
+						*clipboard-providers-clipmethod*
+To integrate with the Vim's actual clipboard functionality, the 'clipmethod'
+option is used on all platforms.  The names of clipboard providers should be
+put inside the option, and if Vim chooses it, then it overrides the "+" and
+"*" registers.  Note that the "+" and "*" will not be saved in the viminfo at
+all.
+
+						*clipboard-providers-define*
+To defined a clipboard provider, the |v:clipproviders| vim variable is used.
+It is a |dict| where each key is the clipboard provider name, and the value is
+another |dict| declaring the 'available", "copy", and "paste" callbacks: >vim
+	let v:clipproviders["myprovider"] = {
+	    \ "available": Available,
+	    \ "paste": {
+	    \     "+": function("Paste"),
+	    \     "*": function("Paste")
+	    \	},
+	    \ "copy": {
+	    \     "+": function("Paste"),
+	    \     "*": function("Paste")
+	    \	}
+	    \ }
+	set clipmethod^=myprovider
+<
+Each callback can either be a name of a function in a string, a |Funcref|, or
+a |lambda| expression.
+
+If a callback except the "available" callback is not provided, then when Vim
+tries to call it, nothing is done.
+
+						*clipboard-providers-textlock*
+In both the "paste" and "copy" callbacks, it is not allowed to change the
+buffer text, see |textlock|.
+
+						*clipboard-providers-available*
+The "available" is an optional callback does not take any arguments and should
+return a |boolean|, which tells Vim if it is available for use.  If it is not,
+then Vim skips over it and tries the next 'clipmethod' item.  If the
+"available" callback is not provided, Vim assumes the provider is always
+available for use (true).
+
+						*clipboard-providers-paste*
+The "paste" callback takes the following arguments in the following order:
+	1. Name of the register being accessed, either "+" or "*".
+
+It should return a |list| or |tuple| containing the following elements in
+order:
+	1. Register type conforming to |setreg()|
+	2. A |list| of strings to return to Vim, each representing a line.
+
+The first element (the register type), may also have a special value of
+"pass".  In that case, Vim does nothing to the register, meaning the current
```

Comment:
that "pass" value is a bit obscure. Wouldn't it be better if the "paste" function  returns a:


> It should return a |list| or |tuple| containing the following elements in
order:
    1. An action, either "pass" or "set"
	2. Register type conforming to |setreg()|
	3. A |list| of strings to return to Vim, each representing a line.